FEXCore/Source/Utils/AllocatorHooks.cpp

```diff
@@ -1,66 +1,206 @@
 using munmap_hook_type = int (*)(void* addr, size_t length);
+
+#ifdef ENABLE_FEX_ALLOCATOR
+typedef void* (*rp_mmap_hook_type)(size_t size, size_t alignment, size_t* offset, size_t* mapped_size);
+typedef void (*rp_munmap_hook_type)(void* address, size_t offset, size_t mapped_size);
+extern "C" rp_mmap_hook_type rp_mmap_hook;
+extern "C" rp_munmap_hook_type rp_munmap_hook;
+
+#ifndef _WIN32
+mmap_hook_type fex_mmap_hook = ::mmap;
+munmap_hook_type fex_munmap_hook = ::munmap;
 #endif
 
-#ifdef ENABLE_JEMALLOC
+// Assume a 64KB page size until told otherwise.
+static rpmalloc_config_t global_config {
+  .page_size = 64 * 1024,
+  // THP causes crashes for some reason.
+  .enable_huge_pages = 0,
+  .disable_decommit = 0,
+  .page_name = "FEXAllocator",
+  .huge_page_name = "FEXAllocator",
+  .unmap_on_finalize = 0,
+};
+
 void* malloc(size_t size) {
-  return ::je_malloc(size);
+  return ::rpmalloc(size);
 }
 void* calloc(size_t n, size_t size) {
-  return ::je_calloc(n, size);
+  return ::rpcalloc(n, size);
 }
 void* memalign(size_t align, size_t s) {
-  return ::je_memalign(align, s);
+  return ::rpmemalign(align, s);
 }
 void* valloc(size_t size) {
-  return ::je_valloc(size);
+  return ::rpaligned_alloc(global_config.page_size, size);
 }
 int posix_memalign(void** r, size_t a, size_t s) {
-  return ::je_posix_memalign(r, a, s);
+  void* ptr;
+  auto res = ::rpposix_memalign(&ptr, a, s);
+  *r = ptr;
+  return res;
 }
 void* realloc(void* ptr, size_t size) {
-  return ::je_realloc(ptr, size);
+  return ::rprealloc(ptr, size);
 }
 void free(void* ptr) {
-  return ::je_free(ptr);
+  return ::rpfree(ptr);
 }
 size_t malloc_usable_size(void* ptr) {
-  return ::je_malloc_usable_size(ptr);
+  return ::rpmalloc_usable_size(ptr);
 }
 void* aligned_alloc(size_t a, size_t s) {
-  return ::je_aligned_alloc(a, s);
+  return ::rpaligned_alloc(a, s);
 }
 void aligned_free(void* ptr) {
-  return ::je_free(ptr);
+  return ::rpfree(ptr);
 }
 
-#ifndef _WIN32
-extern "C" mmap_hook_type je___mmap_hook;
-extern "C" munmap_hook_type je___munmap_hook;
+void InitializeThread() {
+  rpmalloc_thread_initialize();
+}
 
-void SetJemallocMmapHook(mmap_hook_type Hook) {
-  je___mmap_hook = Hook;
+[[nodiscard]]
+constexpr uint64_t AlignUp(uint64_t value, uint64_t size) {
+  return value + (size - value % size) % size;
 }
-void SetJemallocMunmapHook(munmap_hook_type Hook) {
-  je___munmap_hook = Hook;
+
+static void* FEX_rp_mmap(size_t size, size_t alignment, size_t* offset, size_t* mapped_size) {
+#define pointer_offset(ptr, ofs) (void*)((char*)(ptr) + (ptrdiff_t)(ofs))
+  // If the alignment is less than the operating page size then alignment is guaranteed. Just remove it.
+  if (alignment < global_config.page_size) {
+    alignment = 0;
+  }
+
+  size_t map_size = AlignUp(size + alignment, global_config.page_size);
+#ifdef _WIN32
+  auto ptr = VirtualAlloc(0, map_size, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
```

Comment:
Do we want to actually commit here? The explicit commit func suggests not but unsure of rpmalloc internals